### ISSUE TITLE
test_openFileDescriptors in process tests fails sometimes with an extra file descriptor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7633]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7633|
|Type|enhancement|
|Created|2014-09-08 22:44:14Z|
|Branch|https://github.com/twisted/twisted/tree/open-file-descriptors-7633|

See for example [this build](https://buildbot.twistedmatrix.com/builders/natty64-py2.7maint/builds/4160/steps/trial/logs/problems).

<details><summary>Searchable metadata</summary>

```
trac-id__7633 7633
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__branches_open_file_descriptors_7633 branches/open-file-descriptors-7633
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1410216254740359 1410216254740359
changetime__1445217216921098 1445217216921098
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

This might be a duplicate of [#4747](https://github.com/twisted/twisted/issues/4747).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [43181]) Branching to open-file-descriptors-7633.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

I am putting this into the milestone because we are supposed to release with green supported builders, and this is keeping one of the builders red.

I'm also putting it into review, though!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Where do the extra file descriptors that cause the test to fail come from?  The comment in the branch that suggests the platform might interfere with a more straightforward test should give a concrete example of where this happens.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [43185]) explain the cases where file descriptors are left open

refs [#7633](https://github.com/twisted/twisted/issues/7633)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_4):
> Where do the extra file descriptors that cause the test to fail come from?  The comment in the branch that suggests the platform might interfere with a more straightforward test should give a concrete example of where this happens.

As it happens the one that is interfering right now is `/dev/urandom`, but really it's a generalization of the special-case the test already had (the file descriptor left open by `_listOpenFDs`, which _could_ be addressed by the implementation of that function, so isn't really part of what the test is covering).  I think this is a security enhancement made on Python 2.7's development branch.

I put a comment in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Just thinking out loud, I wonder if running the child process with `subprocess.Popen(..., close_fds=True)` first to collect a list of descriptors that are open in a child process launched by someone else's implementation and comparing that list against the descriptors that are open when we use `spawnProcess` (our implementing).

Presumably this still isn't flawless (the obvious drawback is that the test will fail if a bug `subprocess.Popen` causes it to not close some descriptors - though as of now its implementation of this feature is simpler than ours).  Maybe it's still a good trade-off, maybe not.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_7):
> Just thinking out loud, I wonder if running the child process with `subprocess.Popen(..., close_fds=True)` first to collect a list of descriptors that are open in a child process launched by someone else's implementation and comparing that list against the descriptors that are open when we use `spawnProcess` (our implementing).
> 
> Presumably this still isn't flawless (the obvious drawback is that the test will fail if a bug `subprocess.Popen` causes it to not close some descriptors - though as of now its implementation of this feature is simpler than ours).  Maybe it's still a good trade-off, maybe not.

If we gather the expected results this way, there is also the possibility that the code generating these results is just buggy (`_listOpenFDs` returns an empty list, for example).  I think it would be worse than what's in the branch now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [43188]) Since itamar complained. refs [#7633](https://github.com/twisted/twisted/issues/7633)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

r43188 seems like a bad change to me.  Apart from being unrelated to the issue in the ticket description, it changes the failure behavior of the test to be worse.

Consider:

```
>>> eval("oh no some junk")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1
    oh no some junk
        ^
SyntaxError: invalid syntax
```

vs

```
>>> json.loads("oh no some junk")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/json/__init__.py", line 326, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 365, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 383, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```

Which would you rather appear in a test log file you have to read on buildbot to understand a bug that arises?

If this change needs to be part of this branch, please change the test code to preserve visibility of the unparseable output from the child process.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [43189]) Revert "Since itamar complained. refs [#7633](https://github.com/twisted/twisted/issues/7633)"

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_10):
> r43188 seems like a bad change to me.  Apart from being unrelated to the issue in the ticket description, it changes the failure behavior of the test to be worse.

You make a good point, I've reverted it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

I forced a new build so that the buildbot is up to date with these changes.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml commented|
|-|-|

I haven't reviewed for style guide compliance.

1. I think this has one 'itself' too many.
```
 # The call to "os.listdir()" itself that _listOpenFDs's implementation
 # uses itself needs to open a file descriptor (with "opendir"), which
 # shows up in its result.
```
2. Given that the test is still non-deterministic, it's probably a good idea to mention that in the final assertion, pointing readers to the (great!) comment in the test itself.
3. It might be useful to have a summary of the testing strategy near the top: "open an unlikely file descriptor in the parent and verify that it's not open in the child"

Hope this helps.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml set owner to @glyph|
|-|-|

Forgot to do the reassignment step. My review is in comment:14

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml commented|
|-|-|

Please address these minor issues & merge. :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

(In [43288]) Clarify comment refs [#7633](https://github.com/twisted/twisted/issues/7633)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [43289]) More comments as per jml's review, since this test is extremely subtle refs [#7633](https://github.com/twisted/twisted/issues/7633)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set status to `closed`|
|-|-|

(In [43290]) Merge open-file-descriptors-7633: Fix intermittent test.

Author: glyph

Reviewer: jml

Fixes: [#7633](https://github.com/twisted/twisted/issues/7633)

test_openFileDescriptors verified too strictly that no unexpected file
descriptors are open in a subprocess.  Sometimes, "extra" file
descriptors are in fact expected (such as when /dev/urandom is opened by
importing the 'random' module).  This adjusts the test to verify
that *twisted* is doing its job and closing the file descriptors before
the process starts, rather than verifying that the list is exactly what
we expect.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_20"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
